src/components/WalletSelector/AuthenticateWalletPending/AuthenticateWalletPendingGnosisSafe.tsx

```diff
@@ -0,0 +1,138 @@
+import Spacer from 'components/core/Spacer/Spacer';
+import { fetchNonce, loginOrCreateUser } from '../authRequestUtils';
+import {
+  getWalletTypeId,
+  listenForGnosisSignature,
+  signMessageWithContractAccount,
+} from '../walletUtils';
+
+type Props = {
+  pendingWallet: AbstractConnector;
+  userFriendlyWalletName: string;
+  setDetectedError: (error: Web3Error) => void;
+};
+
+function AuthenticateWalletPendingGnosisSafe({
+  pendingWallet,
+  userFriendlyWalletName,
+  setDetectedError,
+}: Props) {
+  const { library, account } = useWeb3React<Web3Provider>();
+
+  const [pendingState, setPendingState] = useState<PendingState>(INITIAL);
+
+  const fetcher = useFetcher();
+  const { logIn } = useAuthActions();
+
+  const attemptAuthentication = useCallback(
+    async (address: string) => {
+      setPendingState(PROMPT_SIGNATURE);
+
+      const { nonce, user_exists: userExists } = await fetchNonce(address, fetcher);
+      const signature = await signMessageWithContractAccount(
+        address,
+        nonce,
+        pendingWallet,
+        library
+      );
+
+      setPendingState(LISTENING_ONCHAIN);
+
+      await listenForGnosisSignature(address, nonce, library);
+
+      const payload = {
+        signature,
+        address,
+        wallet_type: getWalletTypeId(pendingWallet),
+        nonce,
+      };
+
+      const { jwt, userId } = await loginOrCreateUser(userExists, payload, fetcher);
+
+      Mixpanel.trackConnectWallet(userFriendlyWalletName, 'Sign In');
+      logIn({ jwt, userId }, address);
+    },
+    [fetcher, library, logIn, pendingWallet, userFriendlyWalletName]
+  );
+
+  useEffect(() => {
+    async function authenticate() {
+      if (account) {
+        try {
+          await attemptAuthentication(account.toLowerCase());
+        } catch (error: unknown) {
+          if (isWeb3Error(error)) {
+            setDetectedError(error);
+          }
+
+          // Fall back to generic error message
+          if (error instanceof Error) {
+            const web3Error: Web3Error = { code: 'AUTHENTICATION_ERROR', ...error };
+            setDetectedError(web3Error);
+          }
+        }
+      }
+    }
+
+    void authenticate();
+  }, [account, setDetectedError, attemptAuthentication]);
+
+  if (pendingState === PROMPT_SIGNATURE) {
+    return (
+      <div>
+        <TitleMedium>Connect with {userFriendlyWalletName}</TitleMedium>
+        <Spacer height={24} />
+        <BodyRegular color={colors.gray50}>
+          Connecting with Gnosis requires an on chain transaction.
+        </BodyRegular>
+        <Spacer height={8} />
+        <BodyRegular color={colors.gray50}>
+          Follow the prompts in the Gnosis Safe app to submit a transaction to sign the message.
+        </BodyRegular>
+        <Spacer height={32} />
+        <BodyRegular color={colors.gray50}>Do not close this window.</BodyRegular>
+      </div>
+    );
+  }
+
+  if (pendingState === LISTENING_ONCHAIN) {
```

Comment:
this logic seems to be shared with the other file – `AddWalletPendingGnosisSafe.tsx`. is it worth consolidating?

Reply:
moved shared code into a new component `GnosisSafePendingMessage` that contains Gnosis Safe specific screens for the pending state, used for both Auth and Add Wallet flows